geode-core/src/main/java/org/apache/geode/internal/cache/Oplog.java

```diff
@@ -1303,16 +1305,18 @@ void close(DiskRegion dr) {
     // while a krf is being created can not close a region
     lockCompactor();
     try {
-      addUnrecoveredRegion(dr.getId());
-      DiskRegionInfo dri = getDRI(dr);
-      if (dri != null) {
-        long clearCount = dri.clear(null);
-        if (clearCount != 0) {
-          totalLiveCount.addAndGet(-clearCount);
-          // no need to call handleNoLiveValues because we now have an
-          // unrecovered region.
+      if (!isDrfOnly()) {
```

Comment:
If this is a safe change to make to close then it seems like you should also avoid calling lockCompactor if isDrfOnly(). It is called to prevent the compactor from running while the code inside the try is running. Or did you determine that isDrfOnly can only be safely called while the compactor is not running? 
The only concern I have about this new isDrfOnly check here is that it also causes us not to call addUnrecoveredRegion. Most of the places that use the unrecoveredRegion count, use it to figure out if this oplog might have some live values. So for this drfOnly oplog it does not have any live values so those places in the code don't need to worry about this oplog having live values.
But it is also used in checkForRecoverableRegion which will call dri.testAndSetRecovered which will associate the dri with its DiskRegionView. I think we probably still need to do this and if I'm correct then we still need to call addUnrecoveredRegion here even if isDrfOnly() is true.
For those places that do use unrecoveredRegionCount > 0 to indicate that something can't be done; could you in those places also check for isDrfOnly? I'm thinking of hasNoLiveValues() and needsCompaction().

Reply:
I think you have addressed most of my concerns. I'm still worried about the following corner case if we don't add a drf only oplog to the unrecovered region count.
What happens if you have a region that still has live data in an old crf oplog and has only deletes in a newer drf oplog. That newer oplog will be kept alive by the older crf so the newer oplog will be marked as drf only. Then someone calls "Region.close" on that region. This causes the persistent data to be kept on disk but removed from memory. Then that same member turns around and recreates the region it closed. If the close did not remember that drf only oplog had an "unrecovered region" for the drf only one then I'm not sure the recovery will work when the region is recreated. Since it does not reread the disk store files (if, crf, drf, krf) in their entirety (like we do during a normal startup) enough info needs to be kept in memory so that we know that drf only oplog has some info about the region we are recreating. I think this association on the recreate happens in this "checkForRecoverableRegion which will call dri.testAndSetRecovered which will associate the dri with its DiskRegionView". 
Does a test exist for this use case (close + recreate with drf only oplog)? If not if you add one you might find that drf only oplogs are handled in some other way during recreate without them being in the unrecoveredRegionCount.

Reply:
We have created the following test cases (for more information please check comments in test cases):

1. Recovery of single region after cache is restarted (`testCompactorRegionMapDeletedForOnlyDrfOplogAfterCompactionAndRecoveryAfterCacheClosed`)

2. Recovery of single region after region is closed and then recreated (`testCompactorRegionMapDeletedForOnlyDrfOplogAfterCompactionAndRecoveryAfterRegionClose`)

In all cases all data have been successfully recovered from Oplog files, and loaded into the cache memory. It seems that this counter in only used to skip compaction of Oplog file (.crf file) when all region that use Oplog (.crf file) didn't yet recover (e.g. region were closed). All orphaned/only .drf files that reference older .crf Opolg cannot be deleted until older .crf Oplog are deleted. So having `unrecoveredRegionCount` on .crf also prevents deleting any newer .drf file that could contain destroy references to it's entries. We could not find the case were not marking .drf only Oplog as unrecovered during close could cause some issues. Could you please share exact steps for the case that you think it will cause problem and I will try to create it?

During these tests we have found another problem in design base (not introduced by this PR) when region is closed and then recreated to start the recovery. If you inspect this code in close() function you will notice that it doesn't make any sense:

```
void close(DiskRegion dr) {
  // while a krf is being created can not close a region
  lockCompactor();
  try {
    if (!isDrfOnly()) {
      addUnrecoveredRegion(dr.getId());
      DiskRegionInfo dri = getDRI(dr);
      if (dri != null) {
        long clearCount = dri.clear(null);
        if (clearCount != 0) {
          totalLiveCount.addAndGet(-clearCount);
          // no need to call handleNoLiveValues because we now have an
          // unrecovered region.
        }
        regionMap.get().remove(dr.getId(), dri);
      }
    }
  } finally {
    unlockCompactor();
  }
}
private void addUnrecoveredRegion(long drId) {
  DiskRegionInfo dri = getOrCreateDRI(drId);
  if (dri.testAndSetUnrecovered()) {
    unrecoveredRegionCount.incrementAndGet();
  }
}

```
Please notice that `addUnrecoveredRegion()` marks `DiskRegionInfo` object as `unrecovered` and increments counter `unrecoveredRegionCount`. This `DiskRegionInfo` object is contained in `regionMap` structure. Then afterwards it removes `DiskRegionInfo` object (that was previously marked as `unrecovered`) from the `regionMap`. This doesn't make any sense, it updated object and then removed it from map to be garbage collected. As you will see later on this will cause some issues when region is recovered.

Please check this code at recovery:
```

/**
 * For each dri that this oplog has that is currently unrecoverable check to see if a DiskRegion
 * that is recoverable now exists.
 */
void checkForRecoverableRegion(DiskRegionView dr) {
  if (unrecoveredRegionCount.get() > 0) {
    DiskRegionInfo dri = getDRI(dr);
    if (dri != null) {
      if (dri.testAndSetRecovered(dr)) {
        unrecoveredRegionCount.decrementAndGet();
      }
    }
  }
}

```
The problem is that geode will not clear counter `unrecoveredRegionCount` in Oplog objects after recovery is done. This is because `checkForRecoverableRegion` will check `unrecoveredRegionCount` counter and perform `testAndSetRecovered`. The `testAndSetRecovered` will always return false, because non of the `DiskRegionInfo` objects in region map have `unrecovered` flag set to **true** (all object marked as `unrecovered` were deleted by close(), and then they were recreated during recovery.... see **note** below). The problem here is that all Oplogs will be fully recovered with the counter incorrectly indicating `unrecoveredRegionCount>0`. This will later on prevent the compaction of recovered Oplogs (the files that have .crf, .drf and .krf) when they reach compaction threshold.

**Note:** During recovery `regionMap` will be recreated from the Oplog files. Since all `DiskRegionInfo` objects are deleted from `regionMap` during the `close()`, they will be recreated by using function `initRecoveredEntry` during the recovery. All `DiskRegionInfo` will be created with flag `unrecovered` set to false.

Reply:
Thanks for adding the additional tests.
Is the solution to the new problem you discovered to not remove the dri from the regionMap in close? I think this could be a different ticket since it is unrelated to the memory leak.

Reply:
We also agree that this new issue should be handled in a new ticket. We will create it and try to propose the solution. As you already said the solution could be just to not remove the `dri` from `regionMap` in `close()` for Oplog that is not .drf only. 

Thank you for review and approval!